### PR TITLE
REF-#4730: make Indexers immutable

### LINKED ISSUE #4730: STY: indexers should be immutable
e.g. _LocationIndexerBase shouldn't be altered outside of `__init__`

### PR DESCRIPTION
**Pull request by jbrockmendel**
Signed-off-by: Brock Mendel <jbrockmendel@gmail.com>

Make _LocIndexer and _iLocIndexer immutable.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4730
- [ ] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

**Review by pyrito — COMMENTED** [review]
LGTM - there are some docstyle things that are failing but you can ignore them for `__getitem__` and `__setitem__` imo using `# noqa: <CODE>`

**Comment by mvashishtha** [comment]
@jbrockmendel there is one more pydocstyle error: https://github.com/modin-project/modin/runs/7581422257?check_suite_focus=true

**Review by vnlitvinov — COMMENTED** [review]
@jbrockmendel please fix pydocstyle and add a note to release-notes about this PR.
Otherwise looks great!

**Comment by jbrockmendel** [comment]
> add a note to release-notes about this PR

I'll do that for now, but longer term, is there a process to change this policy?  Not all edits merit a release note.

**Comment by pyrito** [comment]
@jbrockmendel we are planning to automate the release notes in the future! Sorry about the inconvenience. We realize it's a pain.

**Comment by jbrockmendel** [comment]
Anything left to do here?

**Review by pyrito — DISMISSED** [review]
LGTM

File: modin/pandas/indexing.py [hunk 1]
```diff
@@ -296,6 +338,10 @@ def __getitem__(self, row_lookup, col_lookup, ndim):
             The global row index to retrieve data from.
         col_lookup : slice(None), range or np.ndarray
             The global col index to retrieve data from.
+        row_multiindex_full_lookup : bool
+        col_multiindex_full_lookup : bool
+        row_scalar : bool
+        col_scalar : bool
```
Comment (vnlitvinov) [inline modin/pandas/indexing.py:344]: some nitpicking - ideally this docstring should explain what parameters are here for, what they do etc.

**Review by vnlitvinov — APPROVED** [review]
Thanks @jbrockmendel, very nice!

File: modin/pandas/indexing.py (hunk 1 above)
Comment (jbrockmendel) [reply modin/pandas/indexing.py]: updated

**Review by pyrito — APPROVED** [review]

File: modin/pandas/indexing.py [hunk 2]
```diff
@@ -296,6 +338,14 @@ def __getitem__(self, row_lookup, col_lookup, ndim):
             The global row index to retrieve data from.
         col_lookup : slice(None), range or np.ndarray
             The global col index to retrieve data from.
+        row_multiindex_full_lookup : bool
+            See _multiindex_possibly_contains_key.__doc__.
+        col_multiindex_full_lookup : bool
+            See _multiindex_possibly_contains_key.__doc__.
```
Comment (vnlitvinov) [inline modin/pandas/indexing.py:344]:
```suggestion
        row_multiindex_full_lookup : bool
            Whether `row_lookup` is potentially an all-levels lookup.
        col_multiindex_full_lookup : bool
            Whether `col_lookup` is potentially an all-levels lookup.
```
?

**Review by vnlitvinov — APPROVED** [review]
LGTM!

File: modin/pandas/indexing.py (hunk 2 above)
Comment (jbrockmendel) [reply modin/pandas/indexing.py:344]: I think the _multiindex_possibly_contains_key is more useful here.

Comment (mvashishtha) [reply modin/pandas/indexing.py:344]: `_multiindex_possibly_contains_key` seems ok to me, too.

**Pull request merged**